include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -779,18 +779,50 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
         nodes.target_phantom = target_phantom;
         UnpackPath(packed_path.begin(), packed_path.end(), nodes, unpacked_path);
 
-        util::Coordinate previous_coordinate = source_phantom.location;
-        util::Coordinate current_coordinate;
+        using util::coordinate_calculation::detail::DEGREE_TO_RAD;
+        using util::coordinate_calculation::detail::EARTH_RADIUS;
+
         double distance = 0;
+        double prev_lat =
+            static_cast<double>(toFloating(source_phantom.location.lat)) * DEGREE_TO_RAD;
+        double prev_lon =
+            static_cast<double>(toFloating(source_phantom.location.lon)) * DEGREE_TO_RAD;
+        double prev_cos = std::cos(prev_lat);
         for (const auto &p : unpacked_path)
         {
-            current_coordinate = facade->GetCoordinateOfNode(p.turn_via_node);
-            distance += util::coordinate_calculation::haversineDistance(previous_coordinate,
-                                                                        current_coordinate);
-            previous_coordinate = current_coordinate;
+            const auto current_coordinate = facade->GetCoordinateOfNode(p.turn_via_node);
+
+            const double current_lat =
+                static_cast<double>(toFloating(current_coordinate.lat)) * DEGREE_TO_RAD;
+            const double current_lon =
+                static_cast<double>(toFloating(current_coordinate.lon)) * DEGREE_TO_RAD;
+            const double current_cos = std::cos(current_lat);
+
+            const double sin_dlon = std::sin((prev_lon - current_lon) / 2.0);
+            const double sin_dlat = std::sin((prev_lat - current_lat) / 2.0);
+
+            const double aharv = sin_dlat * sin_dlat + prev_cos * current_cos * sin_dlon * sin_dlon;
+            const double charv = 2. * std::atan2(std::sqrt(aharv), std::sqrt(1.0 - aharv));
+            distance += EARTH_RADIUS * charv;
+
+            prev_lat = current_lat;
+            prev_lon = current_lon;
+            prev_cos = current_cos;
         }
-        distance += util::coordinate_calculation::haversineDistance(previous_coordinate,
-                                                                    target_phantom.location);
+
+        const double current_lat =
+            static_cast<double>(toFloating(target_phantom.location.lat)) * DEGREE_TO_RAD;
+        const double current_lon =
+            static_cast<double>(toFloating(target_phantom.location.lon)) * DEGREE_TO_RAD;
+        const double current_cos = std::cos(current_lat);
+
+        const double sin_dlon = std::sin((prev_lon - current_lon) / 2.0);
+        const double sin_dlat = std::sin((prev_lat - current_lat) / 2.0);
+
+        const double aharv = sin_dlat * sin_dlat + prev_cos * current_cos * sin_dlon * sin_dlon;
+        const double charv = 2. * std::atan2(std::sqrt(aharv), std::sqrt(1.0 - aharv));
+        distance += EARTH_RADIUS * charv;
+
```

Comment:
And whta math functions could we use instead? We do spend like 10% of our runtime in `sin` / `cos` / `atan`.